File: src/main/java/com/bloodbridge/service/DonationApplicationService.java
package com.bloodbridge.service;

import com.bloodbridge.dto.donationApplication.DonationApplicationCreateRequest;
import com.bloodbridge.dto.donationApplication.DonationApplicationResponse;
import com.bloodbridge.entity.DonationApplication;

import java.util.List;

public interface DonationApplicationService {
    DonationApplicationResponse createDonationApplication(DonationApplicationCreateRequest request);

    DonationApplicationResponse getDonationApplicationById(Long id);

    // TODO: Verify that the authenticated hospital owns this blood request
    // before returning its applications.
    List<DonationApplicationResponse> getApplicationsByBloodRequest(Long bloodRequestId);

    List<DonationApplicationResponse> getApplicationsByDonor(Long donorId);

    // TODO: Verify that the authenticated hospital owns this application's
    // blood request before accepting/rejecting.
    DonationApplicationResponse acceptApplication(Long applicationId);

    DonationApplicationResponse rejectApplication(Long applicationId);

    List<DonationApplicationResponse> getMyApplications();

    void deleteDonationApplication(Long applicationId);
}

File: bloodbridge-frontend/src/services/donationApplicationService.js
import API from "./api"

export const applyForBloodRequest = async (bloodRequestId) => {
    const response = await API.post("/donation-applications", {bloodRequestId});
    return response.data;
};

export const getApplicationsByBloodRequest = async (bloodRequestId) => {
    const response = await API.get(`/donation-applications/blood-requests/${bloodRequestId}`);
    return response.data;
}

export const rejectApplication = async(applicationId) => {
    const response = await API.put(`/donation-applications/${applicationId}/reject`);
    return response.data;
}

export const acceptApplication = async(applicationId) => {
    const response = await API.put(`/donation-applications/${applicationId}/accept`);
    return response.data;
}

export const getApplicationsByDonor = async(donorId) => {
    const response = await API.get(`/donation-applications/applications/${donorId}`);
    return response.data;
}

export const withdrawApplication = async(id) => {
    const response = await API.delete(`/donation-applications/${id}`);
    return response.data;
}

export const completeApplication = async(id) => {
    const response = await API.put(`/donation-applications/${id}/completed`);
    return response.data;
}
